src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -63,68 +76,132 @@ void BatchNorm<InputDataType, OutputDataType>::Reset()
+
+    double nElements = 0.0;
+    if (input.n_elem - size != 0)
+      nElements = 1.0 / (input.n_elem - size + eps);
+
+    // Update running mean and running variance.
+    runningMean = (1 - averageFactor) * runningMean + averageFactor *
+        mean.t();
+    runningVariance = (1 - averageFactor) * runningVariance +
+       input.n_elem * nElements *
+       averageFactor * variance.t();
+  }
+  else
+  {
+    // Normalize the input and scale and shift the output.
+    output = input;
+    arma::cube outputTemp(const_cast<arma::Mat<eT>&>(output).memptr(),
+        input.n_rows / size, size, batchSize, false, false);
+
+    outputTemp.each_slice() -= arma::repmat(runningMean.t(),
+        input.n_rows / size, 1);
+    outputTemp.each_slice() /= arma::sqrt(arma::repmat(runningVariance.t(),
+        input.n_rows / size, 1) + eps);
+    outputTemp.each_slice() %= arma::repmat(gamma.t(),
+        input.n_rows / size, 1);
+    outputTemp.each_slice() += arma::repmat(beta.t(),
+        input.n_rows / size, 1);
   }
 }
 
 template<typename InputDataType, typename OutputDataType>
 template<typename eT>
 void BatchNorm<InputDataType, OutputDataType>::Backward(
-    const arma::Mat<eT>& input, const arma::Mat<eT>& gy, arma::Mat<eT>& g)
+    const arma::Mat<eT>& input,
+    const arma::Mat<eT>& gy,
+    arma::Mat<eT>& g)
 {
   const arma::mat stdInv = 1.0 / arma::sqrt(variance + eps);
 
+  g.set_size(arma::size(input));
+  arma::cube gyTemp(const_cast<arma::Mat<eT>&>(gy).memptr(),
+      input.n_rows / size, size, input.n_cols, false, false);
+  arma::cube gTemp(const_cast<arma::Mat<eT>&>(g).memptr(),
+      input.n_rows / size, size, input.n_cols, false, false);
+
   // Step 1: dl / dxhat.
-  const arma::mat norm = gy.each_col() % gamma;
+  arma::cube norm = gyTemp.each_slice() % arma::repmat(gamma.t(),
+      input.n_rows / size, 1);
 
   // Step 2: sum dl / dxhat * (x - mu) * -0.5 * stdInv^3.
-  const arma::mat var = arma::sum(norm % inputMean, 1) %
-      arma::pow(stdInv, 3.0) * -0.5;
+  arma::mat temp = arma::sum(norm % inputMean, 2);
+  arma::mat vars = temp % arma::repmat(arma::pow(stdInv, 3),
+      input.n_rows / size, 1) * -0.5;
 
-  // Step 4: dl / dxhat * 1 / stdInv + variance * 2 * (x - mu) / m +
+  // Step 3: dl / dxhat * 1 / stdInv + variance * 2 * (x - mu) / m +
   // dl / dmu * 1 / m.
-  g = (norm.each_col() % stdInv) + (inputMean.each_col() %
-      var * 2 / input.n_cols);
+  gTemp = (norm.each_slice() % arma::repmat(stdInv,
+      input.n_rows / size, 1) +
+      (inputMean.each_slice() % vars * 2)) / input.n_cols;
 
-  // Step 3: sum (dl / dxhat * -1 / stdInv) + variance *
+  // Step 4: sum (dl / dxhat * -1 / stdInv) + variance *
   // (sum -2 * (x - mu)) / m.
-  g.each_col() += arma::sum(norm.each_col() % -stdInv, 1) / input.n_cols;
+  arma::mat normTemp = arma::sum(norm.each_slice() %
+      arma::repmat(-stdInv, input.n_rows / size, 1) , 2) /
+      input.n_cols;
+  gTemp.each_slice() += normTemp;
```

Comment:
Hey @kartikdutt18 . 
Its okay with me if we want to go ahead and merge this PR right away, but there is a minor bug in this step 4 of the Backward method() as I have described [here](https://iamshnoo.github.io/blog/week-5/) in detail.

As can be seen from the comment for step 4 itself, there should be 2 terms in the expression for normTemp, where the second term should probably just be the following
```
arma::sum(inputMean.each_slice() % vars * -2) , 2)  / input.n_cols
```

Since this PR is already a major change of the layer, maybe you would like to include this as well within this?
Or I can just create a issue for this separately, which we can work on and resolve later as well.

Reply:
> arma::sum(inputMean.each_slice() % vars * -2) , 2)  / input.n_cols

Isn't gTemp initialized to this?

Reply:
Let me do this, I already have the fast backward PR ready (I can open that, I didn't add it in this one since it gave higher gradients earlier, I will open one PR post dinner).

Reply:
I dont think so.
```
gTemp = g (initially)
```

Then in step 3 : 
```
gTemp = (norm.each_slice() % arma::repmat(stdInv,  input.n_rows / size, 1) +
                (inputMean.each_slice() % vars * 2)) / input.n_cols;
```

Then in step 4:
The summation of the second term from step 3 along with a negative sign has to be added again to gTemp.

Reply:
> gTemp = g (initially)

I think it's initialized to this (from the code)
```
  gTemp = (norm.each_slice() % arma::repmat(stdInv,
       input.n_rows / size, 1) +
       (inputMean.each_slice() % vars * 2)) / input.n_cols;
```

Reply:
> I think it's initialized to this (from the code)

No. That's the gradient from a separate branch in the graph.
As you can see from the pictures of the equations I included above, we need to add the thing I mentioned one again.

Reply:
Hey @iamshnoo can you just check the term you are talking about adding, does it add up to zero?

Reply:
> I already have the fast backward PR ready

Sure, open that one, then we can merge it and totally forget about this one after debugging that properly.
That would be the ideal thing to do as I see it.

Reply:
Hey @iamshnoo can you take a look here and tell me if this is relevant 
https://github.com/mlpack/mlpack/pull/1943 ?

Reply:
> does it add up to zero?

You mean,``` arma::sum(inputMean.each_slice() % vars * -2) , 2)  / input.n_cols == 0``` ?

I haven't checked yet, but I have checked that this calculation matches with 2 existing python implementations from cs231n [here](https://github.com/cthorey/CS231/blob/master/assignment2/cs231n/layers.py#L210) and [here](https://kratzert.github.io/2016/02/12/understanding-the-gradient-flow-through-the-batch-normalization-layer.html) which have already been verified by numerical gradient check functions [here](https://github.com/cthorey/CS231/blob/master/assignment2/cs231n/gradient_check.py). So, it should be correct.

Anyway, the faster  implementation of the BatchNorm which Kartik mentioned above, won't need this term (and many others) anyway, and since he will be putting up a PR on that soon, I guess we can leave this issue as a comment for now till that gets completed.

Reply:
> Hey @iamshnoo can you take a look here and tell me if this is relevant #1943 ?

Yep, thanks for pointing that out. That resolves the doubt I had.
That term becoming zero is also the reason why it won't be needed for the faster implementation, which is what I just figured out :)

Reply:
Awesome, Anyways both of the implementations gave the same value so I thought the above implementation was correct. Opened #2512.

Reply:
Alright so the backward computation we currently have is correct and so, I will resolve this conversation here. 
If you want you can open a new issue to discuss about further optimization to the backward function but IMO it should not be done here because the work for this PR is complete.

Reply:
Right.